### PR TITLE
Fix insertText on selecting the whole inline node

### PR DESCRIPTION
**Pull request by zhujinxuan**
I should fix https://github.com/ianstormtaylor/slate/issues/1494

My solution is that: 
1. after insertText, the selection should be collapsed
2. if selection is collapsed, do not reset inline focus/anchor to the nextText.

### CODE REVIEW

File: packages/slate/src/changes/at-current-range.js [hunk 1]
```diff
@@ -192,6 +192,10 @@ Changes.insertText = (change, text, marks) => {
   marks = marks || selection.marks || document.getInsertMarksAtRange(selection)
   change.insertTextAtRange(selection, text, marks)
 
+  if (selection.isExpanded) {
+    change.select(change.value.selection.collapseToStart(), { snapshot: false })
+  }
+
```
Comment (ianstormtaylor) [inline packages/slate/src/changes/at-current-range.js]: Can you add a test that verifies this logic is working?

File: packages/slate-react/src/plugins/after.js [hunk 2]
```diff
@@ -553,24 +553,27 @@ function AfterPlugin() {
...
-      !anchorInline.isVoid &&
-      anchorOffset == anchorText.text.length
-    ) {
-      const block = document.getClosestBlock(anchorKey)
-      const next = block.getNextText(anchorKey)
-      if (next) range = range.moveAnchorTo(next.key, 0)
-    }
+    if (range.isExpanded) {
+      if (
+        range.isExpanded &&
```
Comment (ianstormtaylor) [inline packages/slate-react/src/plugins/after.js]: Isn't this extra `range.isExpanded` unnecessary with the outer `if`?

File: packages/slate-react/src/plugins/after.js [hunk 3]
```diff
@@ -553,24 +553,27 @@ function AfterPlugin() {
...
-    if (
-      anchorInline &&
-      !anchorInline.isVoid &&
-      anchorOffset == anchorText.text.length
-    ) {
-      const block = document.getClosestBlock(anchorKey)
-      const next = block.getNextText(anchorKey)
-      if (next) range = range.moveAnchorTo(next.key, 0)
-    }
+    if (range.isExpanded) {
```
Comment (ianstormtaylor) [inline packages/slate-react/src/plugins/after.js]: Why only do it when expanded? Can the comment above be updated to mention why this is necessary.

File: packages/slate-react/src/plugins/after.js (hunk 2 above)
Comment (zhujinxuan) [reply packages/slate-react/src/plugins/after.js]: My mistake.  I will delete that.

File: packages/slate-react/src/plugins/after.js (hunk 3 above)
Comment (zhujinxuan) [reply packages/slate-react/src/plugins/after.js]: Comment added.  The reason for `range.isExpanded` is that we want this inline-end normalization works when user is selecting a text.  And we put `collapseToStart` in `insertText` to ensure the text is collapsed.

File: packages/slate/src/changes/at-current-range.js (hunk 1 above)
Comment (zhujinxuan) [reply packages/slate/src/changes/at-current-range.js]: The reason is that, in `insertTextAtRange` or in `deleteAtRange`, the result selection can be 
```
{
anchorKey: anchorText,
anchorOffset: anchorText.text.length,
focusKey: getNextText(anchorText),
focusOffset: 0
}
```
, which is a `range.isExpanded === true`

I have tested it in `:8080/dev.html`.
Shall I add two tests both in `insertTextAtRange` and `insertText` with {normalize: false} to ensure that
1. `insertTextAtRange` renders isExpanded:true
2. `insertText` renders isExpanded:false

Comment (zhujinxuan) [reply packages/slate/src/changes/at-current-range.js]: Test added~

**Comment by ianstormtaylor** [comment]
@zhujinxuan looking over this PR's changes again, and trying to compare it to the discussion in https://github.com/ianstormtaylor/slate/issues/1494, I'm not sure this is following that behavior. If we only ever normalize when the selection is expanded, we end up with an inconsistent behavior. 

I'd be open to a solution of never normalizing, but only if it works cross-browser. Or open to having the inline element removed if it's entirely spanned.

**Comment by zhujinxuan** [comment]
@ianstormtaylor 
Can I know what the inconsistent behavior is?  I mean the inconsistent behavior about cross-browser.

**Comment by zhujinxuan** [comment]
How about that:
1. `range ===findRange(native, value)`
2. if value.selection is almost the range, keep the selection
    otherwise, reselect range with onSelect normalization

**Comment by ianstormtaylor** [comment]
@zhujinxuan last I checked, there was a cross-browser issue in Chrome/Safari where they wouldn't actually let you put the cursor in the end of an inline element if it was adjacent to a text node, or other inline element. They'd automatically move the cursor to the start of the next element. 

So we adopted that behavior in Slate to prevent that from leading to incompatibilities.

**Comment by zhujinxuan** [comment]
@ianstormtaylor Use another method to 
1. Keep onSelect normalization if selection is changed
2. otherwise, use the selection from the slate

The idea is that
1. if selection is changed manually, then the selected range is consistent crossing browser
2. because slate-core is consistent crossing browsers; if the previous manual selection is consistent, then the resulted selection is consistent.

File: packages/slate/test/changes/at-current-range/delete-forward/delete-across-inlines.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+export const output = (
+  <value>
+    <document>
+      <paragraph>
+        <link>
+          <focus />
+        </link>
+        <anchor />two
```
[the remark this answers is not in the document]
Comment (zhujinxuan) [reply packages/slate/test/changes/at-current-range/delete-forward/delete-across-inlines.js]: I am worrying about that too.  I think we shall always have collapseToStart() after the delete()/deleteAtCurrentRange() function.

How do you think about that?

[the remark this answers is not in the document]
Comment (ianstormtaylor) [reply packages/slate/test/changes/at-current-range/delete-forward/delete-across-inlines.js]: That sounds better to me, yup.

[the remark this answers is not in the document]
Comment (zhujinxuan) [reply packages/slate/test/changes/at-current-range/delete-forward/delete-across-inlines.js]: Strange, I see `change.delete` exists `change.collapseToStart`.  But the result selection is expanded.

I am checking whether there is a bug about change.collapseToStart

File: packages/slate-react/src/plugins/after.js [hunk 5]
```diff
@@ -530,6 +530,11 @@ function AfterPlugin() {
     if (!range) return
 
     const { anchorKey, anchorOffset, focusKey, focusOffset } = range
+    if (
```
[the remark this answers is not in the document]
Comment (zhujinxuan) [reply packages/slate-react/src/plugins/after.js]: Fixed~

File: packages/slate-react/src/plugins/after.js [hunk 6]
```diff
@@ -530,6 +530,11 @@ function AfterPlugin() {
...
+      change.value.document.areTwoRangesVisiblyTheSame(value.selection, range)
+    ) {
+      return
+    }
```
[the remark this answers is not in the document]
Comment (zhujinxuan) [reply packages/slate-react/src/plugins/after.js]: Fixed~

File: packages/slate-react/src/plugins/after.js [hunk 7]
```diff
@@ -530,6 +530,11 @@ function AfterPlugin() {
     if (!range) return
 
     const { anchorKey, anchorOffset, focusKey, focusOffset } = range
+    if (
+      change.value.document.areTwoRangesVisiblyTheSame(value.selection, range)
```
[the remark this answers is not in the document]
Comment (zhujinxuan) [reply packages/slate-react/src/plugins/after.js]: Fixed~ My mistake

File: packages/slate/src/models/node.js [hunk 8]
```diff
@@ -224,6 +224,75 @@ class Node {
...
+   * Verify whether the positions are almost the same
+   * @param {String} key1
+   * @param {Number} offset1
+   * @param {String} key2
+   * @param {Number} offset2
+   * @return {Boolean}
+   *
+   * **/
+
+  areTwoCollapsedRangeVisiblyTheSame(range1, range2) {
```
[the remark this answers is not in the document]
Comment (zhujinxuan) [reply packages/slate/src/models/node.js]: Fixed~

File: packages/slate/src/models/node.js [hunk 9]
```diff
@@ -224,6 +224,75 @@ class Node {
...
   }
 
+  /**
+   * Verify whether the anchor and focus of two ranges are visibly the same
+   * @param {Range} range1
+   * @param {Range} range2
+   * @return {Boolean}
+   * **/
+
+  areTwoRangesVisiblyTheSame(range1, range2) {
```
[the remark this answers is not in the document]
Comment (zhujinxuan) [reply packages/slate/src/models/node.js]: Fixed~

File: packages/slate/src/models/node.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (zhujinxuan) [reply packages/slate/src/models/node.js]: Fixed~

File: packages/slate/test/changes/at-current-range/delete-forward/delete-across-inlines.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhujinxuan) [reply packages/slate/test/changes/at-current-range/delete-forward/delete-across-inlines.js]: It seems there exists a bug in slate-hyperscript;
both 
```
<inline><anchor/></inline><focus/>
```
and 
```
<inline><cursor/></inline>
```
works

**Pull request closed without merging**